iostat/iostat.go

```diff
@@ -0,0 +1,328 @@
+
+			if title {
+				// skip the title line
+				title = false
+				continue
+			}
+			if first {
+				//next interval separated by a newline
+				first = false
+				keys = []string{}
+				data = []string{}
+
+				/////////////////////////////////////////////////////////////////////////////
+				// HERE: 	possible parsing output to get timestamp delivered by iostat
+				//			the default format: "MM/DD/YYYY HH:MM:SS AM/PM"
+				/////////////////////////////////////////////////////////////////////////////
+
+				// getting timestamp as current local time
+				// in format: "YYYY-MM-DD HH:MM:SS.fffffffff +/-HHMM TZ
+				iostat.timestamp = time.Now()
+				continue
+			}
+
+			if strings.HasSuffix(line[0], ":") {
+				if len(line) > 1 {
+					statType = strings.ToLower(strings.TrimSuffix(line[0], ":"))
+					statNames = replaceByPerSec(line[1:])
+					continue
+				}
+			}
+
+			if len(statNames) == 0 || len(statType) == 0 {
+				fmt.Fprintf(os.Stderr, "Invalid format of iostat output data\n")
+				break
+			}
+			if len(line) > len(statNames) {
+				// subType is defined
+				statSubType = line[0]
+				line = line[1:]
+			} else {
+				statSubType = ""
+			}
+
+			if len(line) == len(statNames) && len(statNames) != 0 {
+				for i, statName := range statNames {
+
+					if statSubType != "" {
+						key = statType + "/" + statSubType + "/" + statName
+					} else {
+						key = statType + "/" + statName
+					}
+
+					keys = append(keys, key)
+					data = append(data, line[i])
+				}
+			}
+
+			if strings.ToLower(statSubType) == "all" {
+				// all available metrics keys collected
+				first = true // for next scan skip first line
+
+				if len(iostat.keys) == 0 {
+
+					if len(keys) == 0 {
+						panic(errors.New("Error getting iostat metrics namespace"))
+					}
+
+					iostat.mutex.Lock()
+					iostat.keys = make([]string, len(keys))
+					for i, k := range keys {
+						iostat.keys[i] = joinNamespace(createNamespace(k))
+					}
+					iostat.mutex.Unlock()
+				}
+
+				if len(data) != len(iostat.keys) {
+					panic(errors.New("Invalid parsing iostat output"))
+				}
+
+				iostat.mutex.Lock()
+				for i, d := range data {
+					v, err := strconv.ParseFloat(strings.TrimSpace(d), 64)
+					if err == nil {
+						iostat.data[iostat.keys[i]] = v
+					} else {
+						fmt.Fprintln(os.Stderr, "Invalid metric value", err)
+						iostat.data[iostat.keys[i]] = nil
+					}
+
+				}
+				iostat.mutex.Unlock()
+			}
+		} // end of scanning
+	}() // end of go routine
+
+	err = cmd.Start()
+	if err != nil {
+		fmt.Fprintln(os.Stderr, "Error starting iostat", err)
+		return nil, fmt.Errorf(fmt.Sprintf("Error starting iostat command, err=%+v", err))
+	}
+	// we need to wait until we have our metric types
+	st := time.Now()
+	for {
+		iostat.mutex.RLock()
+		c := len(iostat.keys)
+		iostat.mutex.RUnlock()
+		if c > 0 {
+			break
+		}
+		if time.Since(st) > defaultTimeout {
+			return nil, fmt.Errorf("Timed out waiting for metrics from iostat")
+		}
+	}
```

Comment:
New issue was created for that: https://github.com/intelsdi-x/snap-plugin-collector-iostat/issues/4

Pull request resolving this issue: https://github.com/intelsdi-x/snap-plugin-collector-iostat/pull/3